File: inspect_falsepositives.mjs
import fs from 'fs';
import { computeSignals, DEFAULT_THRESHOLD } from './src/detection.js';

const accounts = fs.readFileSync('public/accounts.csv', 'utf8').trim().split('\n').slice(1).map(line => {
  const [account_id, type, category, age_days, is_mule] = line.split(',');
  return { account_id, type, category, age_days: +age_days, is_mule: is_mule === 'true' };
});

const txns = fs.readFileSync('public/transactions.csv', 'utf8').trim().split('\n').slice(1).map(line => {
  const [txn_id, sender_id, receiver_id, amount, timestamp] = line.split(',');
  return {
    txn_id,
    sender_id,
    receiver_id,
    amount: +amount,
    timestamp,
    ts: new Date(timestamp).getTime() / 1000,
  };
});

const scores = computeSignals(accounts, txns);
const above = Array.from(scores.values()).filter(s => s.score >= DEFAULT_THRESHOLD).sort((a, b) => b.score - a.score);
console.log('threshold', DEFAULT_THRESHOLD);
console.log('above threshold count', above.length);
above.slice(0, 50).forEach(s => console.log(s.id, s.score.toFixed(1), s.isMule ? 'mule' : 'legit'));
const legit = above.filter(s => !s.isMule);
console.log('legit above threshold count', legit.length);
console.log('legit top', legit.slice(0, 20).map(s => ({ id: s.id, score: +s.score.toFixed(1) })));

File: inspect_falsepositives.py
import csv
import os
import sys
from datetime import datetime

sys.path.insert(0, os.path.join(os.getcwd(), 'src'))
from detection import computeSignals, DEFAULT_THRESHOLD

accounts = []
with open('public/accounts.csv', newline='', encoding='utf-8') as f:
    reader = csv.DictReader(f)
    for row in reader:
        accounts.append({
            'account_id': row['account_id'],
            'type': row['type'],
            'category': row['category'],
            'age_days': int(row['age_days']),
            'is_mule': row['is_mule'].strip().lower() == 'true'
        })

transactions = []
with open('public/transactions.csv', newline='', encoding='utf-8') as f:
    reader = csv.DictReader(f)
    for row in reader:
        transactions.append({
            'txn_id': row['txn_id'],
            'sender_id': row['sender_id'],
            'receiver_id': row['receiver_id'],
            'amount': float(row['amount']),
            'timestamp': row['timestamp'],
            'ts': datetime.strptime(row['timestamp'], '%Y-%m-%dT%H:%M:%S').timestamp()
        })

scores = computeSignals(accounts, transactions)
above = [s for s in scores.values() if s.score >= DEFAULT_THRESHOLD]
above.sort(key=lambda s: (-s.score, s.id))
print('threshold', DEFAULT_THRESHOLD)
print('above threshold count', len(above))
for s in above[:50]:
    print(s.id, f'{s.score:.1f}', 'mule' if s.isMule else 'legit')
legit = [s for s in above if not s.isMule]
print('legit above threshold count', len(legit))
print('legit top', [(s.id, round(s.score, 1)) for s in legit[:10]])
